cupy/core/carray.cuh

```diff
@@ -226,19 +246,35 @@ private:
   int index_[ndim];
 
 public:
-  __device__ int size() const {
+  __device__ ptrdiff_t size() const {
     return size_;
   }
 
-  __device__ void set(int i) {
-    unsigned int a = i;
-    for (int dim = ndim; --dim > 0; ) {
-      unsigned int s = shape_[dim];
-      index_[dim] = (a % s);
-      a /= s;
+  __device__ void set(ptrdiff_t i) {
+    if (ndim == 0) {
+      index_[0] = i;
+      return;
     }
-    if (ndim > 0) {
-      index_[0] = a;
+    if (size_ > 1LL << 31) {
+      size_t a = i;
```

Comment:
I think signedness of `a` won't affect the behavior. If so, wouldn't it be more intuitive (easier to read) if every integral types are signed?

(But this is not relevant to this PR)